src/libstd/path.rs

```diff
@@ -0,0 +1,2415 @@
+        iter::order::eq(self.components(), other.components())
+    }
+}
+
+impl PathBuf {
+    fn as_mut_vec(&mut self) -> &mut Vec<u8> {
+        unsafe { mem::transmute(self) }
+    }
+
+    /// Allocate an empty `PathBuf`.
+    pub fn new() -> PathBuf {
+        PathBuf { inner: OsString::from_str("") }
+    }
+
+    /// Allocate a `PathBuf` with contents `path`.
+    pub fn from_path(path: &Path) -> PathBuf {
+        PathBuf { inner: path.as_os_str().to_os_string() }
+    }
+
+    /// Allocate a `PathBuf` with contents `path`.
+    pub fn from_str(path: &str) -> PathBuf {
+        PathBuf { inner: OsString::from_str(path) }
+    }
+
+    /// Allocate a `PathBuf` with contents `path`.
+    pub fn from_os_str(path: &OsStr) -> PathBuf {
+        PathBuf { inner: path.to_os_string() }
+    }
+
+    /// Extend `self` with `path`.
+    ///
+    /// If `path` is absolute, it replaces the current path.
+    ///
+    /// On Windows:
+    ///
+    /// * if `path` has a root but no prefix (e.g. `\windows`), it
+    ///   replaces everything except for the prefix (if any) of `self`.
+    /// * if `path` has a prefix but no root, it replaces `self.
+    pub fn push<P: ?Sized>(&mut self, path: &P) where P: AsPath {
+        // in general, a separator is needed if the rightmost byte is not a separator
+        let mut need_sep = self.as_mut_vec().last().map(|c| !is_sep(*c)).unwrap_or(false);
+
+        // in the special case of `C:` on Windows, do *not* add a separator
+        {
+            let comps = self.components();
+            if comps.prefix_len() > 0 &&
+                comps.prefix_len() == comps.path.len() &&
+                comps.prefix.unwrap().is_drive()
+            {
+                need_sep = false
+            }
+        }
+
+        let path = path.as_path();
+
+        // absolute `path` replaces `self`
+        if path.is_absolute() || path.prefix().is_some() {
+            self.as_mut_vec().truncate(0);
+
+        // `path` has a root but no prefix, e.g. `\windows` (Windows only)
+        } else if path.has_root() {
+            let prefix_len = self.components().prefix_remaining();
+            self.as_mut_vec().truncate(prefix_len);
+
+        // `path` is a pure relative path
+        } else if need_sep {
+            self.inner.push_os_str(OsStr::from_str(MAIN_SEP_STR));
+        }
+
+        self.inner.push_os_str(path.as_os_str());
+    }
+
+    /// Truncate `self` to `self.parent()`.
+    ///
+    /// Returns `None` and does nothing if `self.parent()` is `None`.
+    pub fn pop(&mut self) -> bool {
+        match self.parent().map(|p| p.as_u8_slice().len()) {
+            Some(len) => {
+                self.as_mut_vec().truncate(len);
+                true
+            }
+            None => false
+        }
+    }
+
+    /// Updates `self.file_name()` to `file_name`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `self.file()` is `None`, i.e. if there is no file name to update.
```

Comment:
I would personally expect this to work for all paths and just basically call `.push` if it's a root path. Could you elaborate a bit more on why you decided to have this panic?